File: Mrh.Monad/ResultMonad.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Mrh.Monad
{

    public interface IResultMonad<T>
    {
        
    }

    public class ResultSuccess<T> : IResultMonad<T>
    {

        public readonly T Result;
        
        public ResultSuccess(T value)
        {
            this.Result = value;
        }
    }

    public interface IResultError<T> : IResultMonad<T>
    {

        IResultMonad<T2> To<T1, T2>();
    }

    public class ResultError<T> : IResultError<T>
    {

        public readonly List<string> errors;
        
        /// <summary>
        ///     Used to create an error monad.
        /// </summary>
        /// <param name="errors">A human readable list of errors.</param>
        public ResultError(List<string> errors)
        {
            
        }

        public IResultMonad<T2> To<T1, T2>()
        {
            return new ResultError<T2>(this.errors);
        }
    }

    /// <summary>
    ///     The binds for the result monad.
    /// </summary>
    public static class ResultMonad
    {
        /// <summary>
        ///     Bind when the parameters are synchronous.
        /// </summary>
        /// <param name="monad">The monad.</param>
        /// <param name="func">The function</param>
        /// <typeparam name="T1">The initial type.</typeparam>
        /// <typeparam name="T2">The to type.</typeparam>
        /// <returns>The result.</returns>
        /// <exception cref="Exception"> When the types to don't match.</exception>
        public static IResultMonad<T2> Bind<T1, T2>(
            this IResultMonad<T1> monad,
            Func<T1, IResultMonad<T2>> func)
        {
            if (monad is ResultSuccess<T1> success)
            {
                return func(success.Result);
            }

            if (monad is ResultError<T1> error)
            {
                return error.To<T1, T2>();
            }
            throw new UnknownTypeException(monad.GetType(), $"Unable to determine how to bind the monad {monad.GetType()}.");
        }

        public static async Task<IResultMonad<T2>> Bind<T1, T2>(
            this Task<IResultMonad<T1>> monadT,
            Func<T1, Task<IResultMonad<T2>>> func)
        {
            var monad = await monadT.ConfigureAwait(false);
            if (monad is ResultSuccess<T1> success)
            {
                return await func(success.Result).ConfigureAwait(false);
            }

            if (monad is ResultError<T1> error)
            {
                return error.To<T1, T2>();
            }
            throw new UnknownTypeException(monad.GetType(), $"Unable to determine how to bind the monad {monad.GetType()}.");
        }

        public static async Task<IResultMonad<T2>> Bind<T1, T2>(
            this Task<IResultMonad<T1>> monadT,
            Func<T1, IResultMonad<T2>> func)
        {
            var monad = await monadT.ConfigureAwait(false);
            if (monad is ResultSuccess<T1> success)
            {
                return func(success.Result);
            }

            if (monad is ResultError<T1> error)
            {
                return error.To<T1, T2>();
            }
            throw new UnknownTypeException(monad.GetType(), $"Unable to determine how to bind the monad {monad.GetType()}.");
        }
    }
}

File: typescript/webSocket/src/app/share/Api/Monad/ResultMonad.ts
/**
 * Base interface for the result monad.
 */
export interface IResultMonad<T> {
  bind<TB>(func: (value: T) => IResultMonad<TB>): IResultMonad<TB>;

  error(func: (errors: string[]) => void): IResultMonad<T>;

  busy(func: () => void): IResultMonad<T>;

  accessDenied(func: (errors: string[]) => void): IResultMonad<T>;

  allErrors(func: () => void): IResultMonad<T>;
}

/**
 * Used to create a result monad.
 * @param value The value to create.
 */
export function ToResult<T>(value: T): IResultMonad<T> {
  return new ResultSuccess(value);
}

export function ToError<T>(errors: string[]): IResultMonad<T> {
  return new ResultError(errors);
}

export class ResultSuccess<T> implements IResultMonad<T> {
  constructor(
    public readonly value: T
  ) {

  }

  accessDenied(func: (errors: string[]) => void) {
    return this;
  }

  bind<TB>(func: (value: T) => IResultMonad<TB>) {
    return func(this.value);
  }

  busy(func: () => void) {
    return this;
  }

  error(func: (errors: string[]) => void) {
    return this;
  }

  allErrors(func: () => void) {
    return this;
  }
}

export interface IResultError<T> extends IResultMonad<T> {

  to<TB>(): IResultMonad<TB>;
}

export abstract class BaseResultError<T> implements IResultError<T> {

  accessDenied(func: (errors: string[]) => void): IResultMonad<T> {
    return this;
  }

  allErrors(func: () => void): IResultMonad<T> {
    func();
    return this;
  }

  bind<TB>(func: (value: T) => IResultMonad<TB>): IResultMonad<TB> {
    return this.to();
  }

  busy(func: () => void): IResultMonad<T> {
    return this;
  }

  error(func: (errors: string[]) => void): IResultMonad<T> {
    return this;
  }

  to<TB>(): IResultMonad<TB> {
    return undefined;
  }

}

export class ResultError<T> extends BaseResultError<T> {

  constructor(
    public readonly errors: string[]
  ) {
    super();

  }

  error(func: (errors: string[]) => void): IResultMonad<T> {
    func(this.errors);
    return this;
  }

  to<TB>(): IResultMonad<TB> {
    return new ResultError(this.errors);
  }

}

export class ResultAccessDenied<T> extends BaseResultError<T> {

  constructor(
    public readonly errors: string[]
  ) {
    super();
  }

  accessDenied(func: (errors: string[]) => void): IResultMonad<T> {
    func(this.errors);
    return this;
  }

  to<TB>(): IResultMonad<TB> {
    return new ResultAccessDenied(this.errors);
  }
}

export class ResultBusy<T> extends BaseResultError<T> {

  busy(func: () => void): IResultMonad<T> {
    func();
    return this;
  }

  to<TB>(): IResultMonad<TB> {
    return new ResultBusy();
  }
}
